adafruit_irremote.py

```diff
@@ -192,30 +193,56 @@ def decode_bits(self, pulses, debug=False):
                 output[i // 8] |= 1
         return output
 
-    def read_pulses(self, input_pulses, max_pulse=10000, blocking=True):
-        """Read out a burst of pulses until a pulse is longer than ``max_pulse``.
+    def read_pulses_non_blocking(self, input_pulses, max_pulse=10000, pulse_window=0.10):
+        """Read out a burst of pulses without blocking until pulses stop for a specified
+            period (pulse_window), pruning pulses after a pulse longer than ``max_pulse``.
 
-           :param ~pulseio.PulseIn input_pulses: Object to read pulses from
-           :param int max_pulse: Pulse duration to end a burst
-           :param bool blocking: If True, will block until pulses found.
-               If False, will return None if no pulses.
-               Defaults to True for backwards compatibility
+            :param ~pulseio.PulseIn input_pulses: Object to read pulses from
+            :param int max_pulse: Pulse duration to end a burst
+            :param float pulse_window: pulses are collected for this period of time
            """
-        if not input_pulses and not blocking:
-            return None
-        received = []
+        received = None
+        recent_count = 0
+        pruning = False
         while True:
-            while not input_pulses:
-                pass
-            while input_pulses:
+            try:
                 pulse = input_pulses.popleft()
+                recent_count += 1
                 if pulse > max_pulse:
-                    if not received:
+                    if received is None:
                         continue
-                    else:
-                        return received
-                received.append(pulse)
-        return received
+                    pruning = True
+                if not pruning:
+                    if received is None:
+                        received = []
+                    received.append(pulse)
+            except IndexError:
+                if recent_count == 0:
+                    return received
+                recent_count = 0
+                time.sleep(pulse_window)
+
+    # pylint: disable-msg=too-many-arguments
+    def read_pulses(self, input_pulses, max_pulse=10000, blocking=True,
+                    pulse_window=0.10, blocking_delay=0.10):
```

Comment:
After input_pulses add `*, ` to make the remaining kwargs kwarg-only. pylint won't complain about too many then I think. Plus it forces calls to be more readable by labelling the args.

Reply:
I am new to python, so have not been able yet to resolve this. I thought you were saying you could do this:

def _read_pulses_non_blocking(self, input_pulses*, max_pulse=10000, pulse_window=0.10):

but that generates an invalid syntax error. Can you direct me to an explanation somewhere? I will keep looking around.

Reply:
I meant read_pulses() in the previous example. 

The following works, but I'm guessing this is not your suggestion:

def read_pulses(self, *input_pulses, max_pulse=10000, blocking=True,
                pulse_window=0.10, blocking_delay=0.10):

and then call the non-blocking method like this:

pulses = self._read_pulses_non_blocking(input_pulses[0], max_pulse, pulse_window)

Still open to suggestions.

Reply:
Sorry for not replying sooner. What I was thinking is:

`def read_pulses(self, input_pulses, *, max_pulse=10000, blocking=True,
pulse_window=0.10, blocking_delay=0.10):`

This looks like a good overview of it: https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html